pkg/plugin/velero/item_snapshotter/v1/item_snapshotter.go

```diff
@@ -0,0 +1,139 @@
+/*
+Copyright the Velero contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	"context"
+	"time"
+
+	"github.com/vmware-tanzu/velero/pkg/plugin/velero"
+
+	"k8s.io/apimachinery/pkg/runtime"
+
+	api "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+)
+
+type AlsoHandlesInput struct {
+	ItemToSnapshot runtime.Unstructured
+	Backup         *api.Backup
+}
+
+type SnapshotItemInput struct {
+	Item   runtime.Unstructured
+	Params map[string]string
+	Backup *api.Backup
+}
+
+type SnapshotItemOutput struct {
+	UpdatedItem      runtime.Unstructured
+	SnapshotID       string
+	SnapshotMetadata map[string]string			// SnapshotMetadata is information in addition to the SnapshotID that the
+												// plugin wants to store in the backup.  SnapshotMetadata will be passed
+												// back in for DeleteSnapshot and CreateItemFromSnapshot
+	AdditionalItems  []velero.ResourceIdentifier
+	HandledItems     []velero.ResourceIdentifier // Items that were handled by this snapshot that can be ignored
+}
+
+// ProgressInput contains the input parameters for the ItemSnapshotter's Progress function.
+type ProgressInput struct {
+	// ItemID is the id of item that was stored in the backup
+	ItemID velero.ResourceIdentifier
+	// SnapshotID is the snapshot ID returned by ItemSnapshotter
+	SnapshotID string
+	// Backup is the representation of the backup resource processed by Velero.
+	Backup *api.Backup
+}
+
+type ProgressOutput struct {
+	Completed                       bool      // True when the operation has completed, either successfully or with a failure
+	Err                             error     // Set when the operation has failed
+	ItemsCompleted, ItemsToComplete int64     // The number of items that have been completed and the items to complete
+	Started, Updated                time.Time // When the upload was started and when the last update was seen.  Time 0 (time.Unix(0, 0)) is returned if unknown.
+}
+
+type DeleteSnapshotInput struct {
+	SnapshotID       string
+	ItemFromBackup   runtime.Unstructured
```

Comment:
I don't quite understand why `ItemFromBackup` is needed in the DeleteSnapshotInput, is it for cascade deletion?

Reply:
We take it in the existing DeleteItemAction, it is probably useful and we have it available.

`// DeleteItemActionExecuteInput contains the input parameters for the ItemAction's Execute function.
type DeleteItemActionExecuteInput struct {
	// Item is the item taken from the pristine backed up version of resource.
	Item runtime.Unstructured
	// Backup is the representation of the restore resource processed by Velero.
	Backup *velerov1api.Backup
}`